File: services/timer/DelayTimer.cpp
//
// Created by Dylan Meadows on 2020-03-12.
//

#include "DelayTimer.h"
#include "SecondTimer.h"
#include <util/Functions.h>
#include <services/CalibrationService.h>

namespace service::timer::DelayTimer {
    const int CLOSE_THRESHOLD = 167;
    const double UPDATE_FACTOR = 1.0;
    const double CLOSE_UPDATE_FACTOR = 0.75;

    const std::shared_ptr<std::vector<int>>
    createStages(const int targetSecond, const int targetDelay, const int calibration) {
        std::shared_ptr<std::vector<int>> stages = std::make_shared<std::vector<int>>(2);
        (*stages)[0] = createStage1(targetSecond, targetDelay, calibration);
        (*stages)[1] = createStage2(targetDelay, calibration);
        return stages;
    }

    const int createStage1(const int targetSecond, const int targetDelay, const int calibration) {
        return util::functions::toMinimumLength(
            SecondTimer::createStage1(targetSecond, calibration) -
            CalibrationService::toMilliseconds(targetDelay)
        );
    }

    const int createStage2(const int targetDelay, const int calibration) {
        return CalibrationService::toMilliseconds(targetDelay) - calibration;
    }

    int calibrate(const int targetDelay, const int delayHit) {
        const int delta = CalibrationService::toMilliseconds(delayHit) -
                          CalibrationService::toMilliseconds(targetDelay);
        if (std::abs(delta) <= CLOSE_THRESHOLD) {
            return static_cast<int>(CLOSE_UPDATE_FACTOR * delta);
        } else {
            return static_cast<int>(UPDATE_FACTOR * delta);
        }
    }
}

File: src/timers/delayTimer.ts
import { toMinimumLength } from '../utils/constants';
import { CalibratorSettings, toMilliseconds } from './calibrator';
import { createSecondPhases } from './secondTimer';

const CLOSE_THRESHOLD = 167;
const UPDATE_FACTOR = 1.0;
const CLOSE_UPDATE_FACTOR = 0.75;

export function createDelayPhases(
  settings: CalibratorSettings,
  targetDelay: number,
  targetSecond: number,
  calibration: number,
): number[] {
  const phase1 = toMinimumLength(
    createSecondPhases(targetSecond, calibration)[0] - toMilliseconds(settings, targetDelay),
  );
  const phase2 = toMilliseconds(settings, targetDelay) - calibration;
  return [phase1, phase2];
}

export function calibrateDelay(
  settings: CalibratorSettings,
  targetDelay: number,
  delayHit: number,
): number {
  const delta = toMilliseconds(settings, delayHit) - toMilliseconds(settings, targetDelay);
  if (Math.abs(delta) <= CLOSE_THRESHOLD) {
    return CLOSE_UPDATE_FACTOR * delta;
  }
  return UPDATE_FACTOR * delta;
}
